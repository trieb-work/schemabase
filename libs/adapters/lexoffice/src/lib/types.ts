// category ID: 8f8664a8-fd86-11e1-a21f-0800200c9a66 = Warenverk√§ufe
export type LexOfficeVoucherItem = {
  amount: number
  taxAmount: number
  taxRatePercent: number
  categoryId: "8f8664a8-fd86-11e1-a21f-0800200c9a66"
}
export type VoucherStatus =
  | "draft"
  | "open"
  | "paid"
  | "paidoff"
  | "voided"
  | "overdue"
  | "accepted"
  | "rejected"
export type LexOfficeVoucher = {
  id: string
  voucherStatus: VoucherStatus
  voucherNumber: string
  totalGrossAmount: number
  files: []
}

export type LexofficeInvoiceObject = {
  type: "salesinvoice"
  voucherNumber: string
  voucherDate: string
  shippingDate?: string
  remark?: string
  voucherStatus: "open" | "paid" | "paidoff" | "voided" | "transferred" | "sepadebit"
  dueDate: string
  totalGrossAmount: number
  totalTaxAmount: number
  taxType: "gross"
  useCollectiveContact: boolean
  contactId?: string
  voucherItems: LexOfficeVoucherItem[]
  version: number
}
