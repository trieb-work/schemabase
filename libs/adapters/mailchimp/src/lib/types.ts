export type ECommerceCustomer = {
  id: string;
  email_address?: string;
  first_name?: string;
  last_name?: string;
  opt_in_status?: boolean;
  company?: string;
};

export type MailchimpMemberStatus =
  | "subscribed"
  | "unsubscribed"
  | "cleaned"
  | "pending"
  | "transactional";
export type Orderline = {
  id: string;
  product_id: string;
  product_variant_id: string;
  quantity: number;
  price: number;
  discount?: number;
  title?: string;
};
export type Promos = { code: string; amount_discounted: number; type: "fixed" };

export type MailchimpOrder = {
  id: string;
  customer: unknown;
  currency_code: string;
  order_total: number | string;
  shipping_total: number | string;
  tax_total: number | string;
  processed_at_foreign: string;
  lines: Orderline[];
  shipping_address: {
    name: string;
    address1: string;
    address2: string;
    city: string;
    postal_code: string;
    country?: string;
    country_code?: string;
    company: string;
  };
  billing_address: {
    name: string;
    address1: string;
    address2: string;
    city: string;
    postal_code: string;
    country?: string;
    country_code?: string;
    company: string;
  };
  discount_total: number;
  promos: Promos[];
  landing_site: string;
  financial_status: "pending";
  fulfillment_status: "pending";
  order_url: string;
};

export type ProductVariant = {
  id: string;
  title: string;
  url: "";
  sku: "pf-leb-5-gemischt";
  price: number;
  inventory_quantity: 0;
  image_url: "https://s3.triebwork.com/static/products/3-lebkuchen-in-bio-folie.jpg";
  backorders: "";
  visibility: "";
  created_at: "2021-02-13T11:37:28+00:00";
  updated_at: "2021-02-13T11:37:29+00:00";
};
