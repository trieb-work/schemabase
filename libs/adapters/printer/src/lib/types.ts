export type Rotate = 0 | 90 | 180;
export type Options = {
  fit_to_page?: boolean;
  rotate?: Rotate;
  paper?: string;
};

export type PrintRequest = {
  /**
   * Base64 encoded content to print
   */
  base64String: string;
  fileName: string;
  printerId: number;
  copies?: number;
  idempotencyKey?: string;
  rotate: Rotate;
  fitToPage?: boolean;
  paper?: string;
};

export type Device = {
  capabilities: {
    bins: ["Automatisch ausw√§hlen"];
    collate: true;
    color: false;
    copies: 9999;
    dpis: ["300x300", "300x600"];
    duplex: boolean;
    extent: [[127, 222], [618, 27093]];
    medias: ["Default"];
    nup: [1, 2, 4, 6, 9, 16];
    papers: {
      "11351 Jewelry Label": [541, 222];
      "11352 Return Address Int": [254, 540];
      "11353 Multi-Purpose": [254, 254];
      "11354 Multi-Purpose": [571, 317];
      "11355 Multi-Purpose": [190, 508];
      "11356 White Name badge": [413, 889];
      "14681 CD/DVD Label": [586, 660];
      "30252 Address": [278, 889];
      "30253 Address (2 up)": [586, 889];
      "30256 Shipping": [587, 1016];
      "30258 Diskette": [540, 698];
      "30277 File Folder (2 up)": [286, 872];
      "30299 Jewelry Label (2 up)": [541, 222];
      "30320 Address": [278, 889];
      "30321 Large Address": [357, 885];
      "30323 Shipping": [540, 1008];
      "30324 Diskette": [540, 698];
      "30325 Video Spine": [190, 1492];
      "30326 Video Top": [460, 778];
      "30327 File Folder": [198, 872];
      "30330 Return Address": [190, 508];
      "30332 1 in x 1 in": [254, 254];
      "30333 1/2 in x 1 in (2 up)": [254, 254];
      "30334 2-1/4 in x 1-1/4 in": [571, 317];
      "30335 1/2 in x 1/2 in (4 up)": [257, 301];
      "30336 1 in x 2-1/8 in": [254, 540];
      "30337 Audio Cassette": [413, 889];
      "30339 8mm Video (2 up)": [190, 714];
      "30345 3/4 in x 2-1/2 in": [190, 635];
      "30346 1/2 in x 1-7/8 in": [127, 476];
      "30347 1 in x 1-1/2 in": [254, 381];
      "30348 9/10 in x 1-1/4 in": [228, 317];
      "30364 Name Badge Label": [587, 1016];
      "30365 Name Badge Card": [589, 889];
      "30370 Zip Disk": [508, 595];
      "30373 Price Tag Label": [248, 508];
      "30374 Appointment Card": [508, 889];
      "30376 Hanging File Insert": [279, 508];
      "30383 PC Postage 3-Part": [571, 1778];
      "30384 PC Postage 2-Part": [587, 1905];
      "30387 PC Postage EPS": [587, 2667];
      "30854 CD Label": [586, 660];
      "30856 Badge Card Label": [618, 1028];
      "30857 Badge Label": [587, 1016];
      "30886 CD Label": [392, 444];
      "99010 Standard Address": [278, 889];
      "99012 Large Address": [357, 885];
      "99014 Name Badge Label": [540, 1008];
      "99014 Shipping": [540, 1008];
      "99015 Diskette": [540, 698];
      "99016 Video Spine": [220, 1476];
      "99016 Video Top": [490, 778];
      "99017 Suspension File": [127, 508];
      "99018 Small Lever Arch": [380, 1899];
      "99019 Large Lever Arch": [587, 1899];
      Banner: [540, 27093];
      "Continuous, Wide": [540, 2794];
    };
    printrate: null;
    supports_custom_paper_size: false;
  };
  computer: {
    createTimestamp: "2020-09-30T06:49:53.454Z";
    hostname: string;
    id: 323632;
    inet: "10.10.6.10";
    inet6: null;
    jre: null;
    name: "DESKTOP-PDTJ4MC";
    state: "connected";
    version: "4.24.0";
  };
  createTimestamp: "2020-09-30T06:49:53.523Z";
  default: false;
  description: string;
  id: number;
  name: string;
  state: "online";
};
