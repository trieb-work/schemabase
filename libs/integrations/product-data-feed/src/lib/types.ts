export type Product = {
  id: string;
  title: string;
  description: string;
  rich_text_description: string;
  image_link?: string;
  additional_image_link?: string;
  link: string;
  price: string;
  sale_price: string;
  condition: string;
  gtin?: string;
  brand?: string;
  unit_pricing_measure?: string;
  availability: "out of stock" | "in stock";
  google_product_category?: string;
};

export type FeedVariant = "googlemerchant" | "facebookcommerce";
