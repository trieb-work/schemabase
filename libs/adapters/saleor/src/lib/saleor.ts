export function adaptersSaleor(): string {
  return "adapters-saleor";
}
