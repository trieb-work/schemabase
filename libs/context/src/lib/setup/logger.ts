import { ContextMissingFieldError, ExtendContextFn } from "../context"
import { createLogger } from "@eci/util/apm"
import winston from "winston"
export type Logger = winston.Logger

/**
 * Fetch the client's configuration and expose it to the context
 */
export const getLogger = (): ExtendContextFn<"logger"> => async (ctx) => {
  if (!ctx.elasticSearch) {
    throw new ContextMissingFieldError("elasticSearch")
  }

  ctx.logger = createLogger(ctx.elasticSearch.loggingServer)

  return ctx
}
