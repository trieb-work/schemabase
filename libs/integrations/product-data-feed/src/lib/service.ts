import ObjectsToCsv from "objects-to-csv";
import { htmlToText } from "html-to-text";
import { Product as RawProduct } from "@eci/adapters/saleor";
// @ts-expect-error it doesn't detec types for some reason
import edjsHTML from "editorjs-html";
import { generateUnitPrice } from "./generate-unit-price";
import { FeedVariant, Product } from "./types";

export interface ProductDataFeedService {
  generateCSV: (
    storefrontProductUrl: string,
    feedVariant: FeedVariant,
    channelSlug: string,
  ) => Promise<string>;
}

export type ProductDataFeedServiceConfig = {
  saleorClient: {
    getProducts: (variables: {
      first: number;
      channel: string;
    }) => Promise<RawProduct[]>;
  };
  channelSlug: string;
};

/**
 * Generate product data as .csv
 */

export class ProductDataFeedGenerator implements ProductDataFeedService {
  public readonly saleorClient: {
    getProducts: (variables: {
      first: number;
      channel: string;
    }) => Promise<RawProduct[]>;
  };
  public readonly channelSlug: string;

  public constructor(config: ProductDataFeedServiceConfig) {
    this.saleorClient = config.saleorClient;
    this.channelSlug = config.channelSlug;
  }

  public async generateCSV(
    storefrontProductUrl: string,
    feedVariant: FeedVariant,
  ): Promise<string> {
    const products = await this.generate(storefrontProductUrl, feedVariant);
    const csv = new ObjectsToCsv(products);
    return await csv.toString();
  }

  private async generate(
    storefrontProductUrl: string,
    feedVariant: FeedVariant,
  ): Promise<Product[]> {
    const rawProducts = await this.saleorClient.getProducts({
      first: 100,
      channel: this.channelSlug,
    });
    const products: Product[] = [];

    for (const rawProduct of rawProducts) {
      // we get the brand from a product attribute called brand
      const brand = rawProduct.attributes.find(
        (x) => x.attribute.name === "brand",
      )?.values[0]?.name;
      const googleProductCategory = rawProduct.attributes.find(
        (x) => x.attribute.name === "googleProductCategory",
      )?.values[0]?.name;

      // if we want to prefer the title instead of the seoTitle
      // const title = product.name ? product.name : product.seoTitle;

      const title = rawProduct.seoTitle ? rawProduct.seoTitle : rawProduct.name;

      let description = "";
      try {
        description = rawProduct.descriptionJson
          ? edjsHTML().parse(JSON.parse(rawProduct.descriptionJson))
          : rawProduct.seoDescription;
      } catch (err) {
        // console.warn(err)
      }

      description =
        feedVariant == "facebookcommerce"
          ? htmlToText(description)
          : description;

      const { hasVariants } = rawProduct.productType;

      if (!rawProduct.variants) {
        continue;
      }

      for (const variant of rawProduct.variants) {
        if (!variant) {
          continue;
        }

        const gtin = hasVariants
          ? variant.metadata?.find((x) => x?.key === "EAN")?.value
          : rawProduct.metadata?.find((x) => x?.key === "EAN")?.value;
        const unit_pricing_measure =
          variant.weight && rawProduct.weight
            ? generateUnitPrice(variant.weight, rawProduct.weight)
            : undefined;
        const product: Product = {
          id: variant.sku,
          title: hasVariants ? `${title} (${variant.name})` : title,
          description,
          rich_text_description:
            feedVariant === "facebookcommerce" ? description : undefined,
          image_link: hasVariants
            ? variant.images && variant.images.length > 0
              ? variant.images[0]?.url
              : ""
            : rawProduct.images && rawProduct.images.length > 0
            ? rawProduct.images[1]?.url
            : "",
          additional_image_link: hasVariants
            ? variant.images?.[1]?.url
            : rawProduct.images?.[2]?.url,
          link: storefrontProductUrl + rawProduct.slug,
          price: `${variant?.pricing?.priceUndiscounted?.gross.amount} ${variant?.pricing?.priceUndiscounted?.gross.currency}`,
          sale_price: `${variant?.pricing?.price?.gross.amount} ${variant.pricing?.price?.gross.currency}`,
          condition: "new",
          gtin,
          brand: brand ?? undefined,
          unit_pricing_measure,
          availability:
            variant.quantityAvailable < 1 || !rawProduct.isAvailableForPurchase
              ? "out of stock"
              : "in stock",
          google_product_category: googleProductCategory ?? undefined,
        };

        products.push(product);
      }
    }
    return products;
  }
}
