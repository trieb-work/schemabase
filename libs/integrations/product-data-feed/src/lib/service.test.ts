import { ProductDataFeedGenerator } from "./service";
import { SaleorClient, WeightUnitsEnum } from "@eci/adapters/saleor";
import { FeedVariant } from "./types";
import { NoopLogger } from "@eci/util/logger";

beforeEach(() => {
  jest.clearAllMocks();
});

describe("generate", () => {
  const mockedSaleorClient = {
    products: async (variables: { first: number; channel: string }) =>
      Promise.resolve({
        products: {
          edges: [
            {
              node: {
                __typename: "Product",
                name: "Name",
                slug: "slug",
                seoDescription: "SeoDescription",
                productType: {
                  hasVariants: true,
                  __typename: "ProductType",
                  name: "name",
                  id: "id",
                },
                channel: variables.channel,
                variants: [
                  {
                    metadata: [],
                    id: "id",
                    name: "name",
                    sku: "sku",
                    quantityAvailable: 5,
                    weight: {
                      unit: WeightUnitsEnum.Kg,
                      value: 2,
                    },
                  },
                ],
                attributes: [],
                metadata: [],
              },
            },
          ],
        },
      }),
  } as unknown as SaleorClient;
  const generator = new ProductDataFeedGenerator({
    saleorClient: mockedSaleorClient,
    channelSlug: "doesn't matter here",
    logger: new NoopLogger(),
  });
  const variants: FeedVariant[] = ["facebookcommerce", "googlemerchant"];
  for (const variant of variants) {
    it("converts the products correctly", async () => {
      const csv = await generator.generateCSV("abc", variant);

      expect(csv).toMatchSnapshot();
    });
  }
});
