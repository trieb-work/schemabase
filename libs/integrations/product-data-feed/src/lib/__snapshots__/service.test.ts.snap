// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate converts the products correctly 1`] = `
"id,title,description,rich_text_description,image_link,additional_image_link,link,price,sale_price,condition,gtin,brand,unit_pricing_measure,availability,google_product_category
sku,Name (name),,,,,abcslug,undefined undefined,undefined undefined,new,,,,out of stock,
"
`;
